networkx/algorithms/community/centrality.py

```diff
@@ -169,3 +215,148 @@ def _without_most_central_edges(G, most_valuable_edge):
+                candidates[sg] = (candidate_edge, centrality)
+            # Otherwise, read from cache
+            else:
+                candidate_edge, centrality = candidates[sg]
+
+            # Update globally most valuable edge and the affected subgraph
+            if optimal_centrality is None or centrality > optimal_centrality:
+                optimal_centrality = centrality
+                global_optimal_edge = candidate_edge
+                affected_subgraph = sg
+
+        # Select globally most valuable edge, remove edge from main graph and component
+        G.remove_edge(*global_optimal_edge)
+        affected_subgraph.remove_edge(*global_optimal_edge)
+
+        # Set the cache to None, such that the candidate edge and centrality will be recomputed for this subgraph in the next iteration
+        candidates[affected_subgraph] = None
+
+        # Check if removing the global_optimal_edge results in the affected subgraph being split
+        new_comps = list(nx.connected_components(affected_subgraph))
+        if len(new_comps) > 1:
+            # Update cache: replace the affected (split) subgraph with the newly generated subgraphs
+            del candidates[affected_subgraph]
+
+            for new_comp in new_comps:
+                candidates[G.subgraph(new_comp).copy()] = None
+
+            return tuple(map(set, [sg.nodes() for sg in candidates]))
+
+
+def _without_most_central_edges_betweenness(
+    G, edge_betweenness, edge_sources, node_contributions
+):
+    original_num_components = nx.number_connected_components(G)
+    num_new_components = original_num_components
+
+    # Cut edge until the component is split
+    while num_new_components <= original_num_components:
+        # Select edge to remove with largest betweenness centrality, do tie-breaking with the node numbers
+        edge_to_remove = max(
+            edge_betweenness,
+            key=edge_betweenness.get,
+        )
+
+        # The new components after removing edge
+        new_components = _update_after_removal(
+            G, edge_to_remove, edge_betweenness, edge_sources, node_contributions
+        )
+        num_new_components = len(new_components)
+
+    return new_components
+
+
+def _update_after_removal(
+    G, edge_to_remove, edge_betweenness, edge_sources, node_contributions
+):
+    """Update betweenness after edge removal using cached contributions."""
+    G.remove_edge(*edge_to_remove)
+
+    affected_nodes = edge_sources[edge_to_remove]
+    del edge_betweenness[edge_to_remove]
+    del edge_sources[edge_to_remove]
+
+    for node in affected_nodes:
+        # Remove old contribution
+        for edge, score in node_contributions[node].items():
+            if edge == edge_to_remove:
+                continue
+            edge_betweenness[edge] -= score
+            edge_sources[edge].remove(node)
+
+        # Recompute shortest paths and contributions
+        S, P, sigma, _ = _single_source_shortest_path_basic(G, node)
```

Comment:
I was thinking something like this to compute a component was split
```suggestion
    component_split = False
    for node in affected_nodes:
        # Remove old contribution
        for edge, score in node_contributions[node].items():
            if edge == edge_to_remove:
                continue
            edge_betweenness[edge] -= score
            edge_sources[edge].remove(node)

        # Recompute shortest paths and contributions
        S, P, sigma, D = _single_source_shortest_path_basic(G, node)
        if edge_to_remove[0] not in D or edge_to_remove[1] not in D:
            component_split = True
```

Reply:
Makes sense!

Reply:
The latest code caches the latest connected components, and only updates the connected components if `component_split is True`